### PR TITLE
Chapter1/homework/5

### PR DESCRIPTION
**Pull request by TelefunkenU87**
Completed homework 5 for chapter 1

### CODE REVIEW

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 1]
```diff
@@ -63,8 +111,14 @@ public static int[] RemoveAt(int[] array, int index)
...
         public static int[] InsertFirst(int[] array, int number)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null || array.Length == 0)
+            {
+                return array;
+            }
+
+            int[] tempArr = new int[array.Length + 1];
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove this useless assignment to local variable 'tempArr'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368&bid=16697853)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 2]
```diff
@@ -29,8 +50,14 @@ public static void Reverse(int[] array)
...
         public static int[] RemoveLast(int[] array)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null || array.Length == 0)
+            {
+                return array;
+            }
+
+            int[] tempArr = new int[array.Length - 1];
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove this useless assignment to local variable 'tempArr'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 3]
```diff
@@ -88,8 +148,28 @@ public static int[] InsertLast(int[] array, int number)
...
+            }
+
+            int[] tempArr = new int[array.Length + 1];
+            int oldArrayIndex = 0;
+            for (int i = 0; i < tempArr.Length; i++)
+            {
+                if (i == index)
+                {
+                    tempArr[i] = number;
+                    i++;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'i' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 1 above)
Comment (TelefunkenU87) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: What would be a good alternative to this variable assignment? The intention is to create a new array that is one larger then the on being added to.

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 3 above)
Comment (TelefunkenU87) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: What would be a good alternative?

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 1 above)
Comment (ellmaster) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: Remove this line completely and put int[] in front of the next line. Same goes for the other remarks. Think about why that works and why your line here is doing nothing. :)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 4]
```diff
@@ -39,8 +66,14 @@ public static int[] RemoveLast(int[] array)
...
         public static int[] RemoveFirst(int[] array)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null || array.Length == 0)
+            {
+                return array;
+            }
+
+            int[] tempArr = new int[array.Length - 1];
```
Comment (ellmaster) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Same.

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 5]
```diff
@@ -75,8 +129,14 @@ public static int[] InsertFirst(int[] array, int number)
...
         public static int[] InsertLast(int[] array, int number)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null || array.Length == 0)
+            {
+                return array;
+            }
+
+            int[] tempArr = new int[array.Length + 1];
```
Comment (ellmaster) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Same.

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 3 above)
Comment (ellmaster) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: I would say in this case it's fine but in more complex scenarios it will lead to confusion. It's better to avoid this. Maybe codacy will be happy if you use index  + 1 explicitly.

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 6]
```diff
@@ -88,8 +148,28 @@ public static int[] InsertLast(int[] array, int number)
...
+            int oldArrayIndex = 0;
+            for (int i = 0; i < tempArr.Length; i++)
+            {
+                if (i == index)
+                {
+                    tempArr[i] = number;
+                    i++;
+                    if (index == array.Length)
+                    {
```
Comment (ellmaster) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: This should be in front of the for loop, shouldn't it?

Comment (TelefunkenU87) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: If you are talking about the if statement that returns the array inside the if statement that checks if the new array is at the requested index this is my reasoning for doing it. 
This check if the spot to insert was at the end of the array. If it is then there is no reason to proceed and just returns there. If it isn't then there would be more of the array to shift over after the insert. If what you are pointing put here is that second if statement.

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368&bid=16697853)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 7]
```diff
@@ -88,8 +144,28 @@ public static int[] InsertLast(int[] array, int number)
...
+            }
+
+            int[] tempArr = new int[array.Length + 1];
+            int oldArrayIndex = 0;
+            for (int i = 0; i < tempArr.Length; i++)
+            {
+                if (i == index)
+                {
+                    tempArr[i] = number;
+                    i++;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'i' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

File: Src/BootCamp.Chapter1/Program.cs [hunk 8]
```diff
@@ -1,9 +1,12 @@
-﻿namespace BootCamp.Chapter1
+﻿using System;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/Program.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

**Comment by TelefunkenU87** [comment]
I removed the code that you requested and played around with the functions to see how it behaved after. I think I see now how it was useless. My original thinking going in was that with C# you needed to declare the size of the array before you could use it. If I am understanding this correctly, if I just declare an array and then assign it to another array it will take the size of the one I am assigning it to? Is assign the right word? saying int[] newArr = FunctionThatReturnsArr()?

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368&bid=16697853)

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 7 above)
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'i' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

**Comment by Almantask** [comment]
Tests are failing, please take a look at the checks (all the way at the bottom) and try to cover all the test cases. (keep in mind edge cases)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368&bid=16697853)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 9]
```diff
@@ -88,8 +149,28 @@ public static int[] InsertLast(int[] array, int number)
...
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'i' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

File: Src/BootCamp.Chapter1/Program.cs [hunk 10]
```diff
@@ -1,9 +1,13 @@
...
+
+namespace BootCamp.Chapter1
 {
     class Program
     {
         static void Main(string[] args)
         {
+            int[] testArr = new int[2] { 0, 1 };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/Program.cs]: Codacy Issue found: [Remove the array size specification; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368&bid=16697853)

File: Src/BootCamp.Chapter1/Program.cs [hunk 11]
```diff
@@ -1,9 +1,13 @@
-﻿namespace BootCamp.Chapter1
+﻿using System;
+
+namespace BootCamp.Chapter1
 {
     class Program
     {
         static void Main(string[] args)
         {
+            int[] testArr = new int[] { 1, 1, 1 };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/Program.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 12]
```diff
@@ -88,8 +154,28 @@ public static int[] InsertLast(int[] array, int number)
...
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'i' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368&bid=16697853)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 13]
```diff
@@ -88,8 +139,29 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns input array.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null || array.Length == 0)
+            {
+                int[] newArr = new int[] { number };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 14]
```diff
@@ -88,8 +139,29 @@ public static int[] InsertLast(int[] array, int number)
...
+            }
+
+            int[] tempArr = new int[array.Length + 1];
+            int oldArrayIndex = 0;
+            for (int i = 0; i < tempArr.Length; i++)
+            {
+                if (i == index)
+                {
+                    tempArr[i] = number;
+                    i++;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'i' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

**Comment by TelefunkenU87** [comment]
Sorry for all the pushes. I'm going to be honest with you. I really don't understand how to read these tests.

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368&bid=16697853)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 15]
```diff
@@ -88,8 +139,33 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns input array.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (index < 0 || array.Length + 1 < index)
+            {
+                return array;
+            }
+            if (array == null || array.Length == 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Change this condition so that it does not always evaluate to 'false'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 16]
```diff
@@ -88,8 +139,33 @@ public static int[] InsertLast(int[] array, int number)
...
+            {
+                int[] newArr = new int[] { number };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 17]
```diff
@@ -88,8 +139,33 @@ public static int[] InsertLast(int[] array, int number)
...
+            }
+
+            int[] tempArr = new int[array.Length + 1];
+            int oldArrayIndex = 0;
+            for (int i = 0; i < tempArr.Length; i++)
+            {
+                if (i == index)
+                {
+                    tempArr[i] = number;
+                    i++;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'i' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

**Comment by TelefunkenU87** [comment]
I think I am getting hung up on implementing this requirement.
"If an array is empty or null, returns input array."
I don't think I fully understand what is being asked and this inability to fully understand it is showing in the tests. I first assumed that "If an array is empty or null, returns input array." was saying that if the array is empty or null just return the array, don't do anything to it. But then I fail tests that are inputting a null or empty array and expecting an array with one element to return. I changed the code to facilitate that and I am now failing tests that are imputing a null or empty array and expecting the same empty array to come back empty.

I'm sure that this failing is rooted in my failure to understand what is being requested. I'll have to walk away, come back to it later, and see if I can understand what is being asked.

I feel like I don't really understand the error messages from the tests and at this point I am unfortunately just throwing code at it hoping they go away. I apologize about all the pushes I made in the last couple hours. I was hoping that by changing code, that I could see the differences in the failed tests and hopefully better understand what was going on with them.

**Comment by Almantask** [comment]
@TelefunkenU87 ping me on discord and we can go on a call.

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368&bid=16697853)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 18]
```diff
@@ -88,8 +151,33 @@ public static int[] InsertLast(int[] array, int number)
...
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'i' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5075368)

**Comment by TelefunkenU87** [comment]
Thank you Kai. I'll ping you soon and see what time works for you.

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 19]
```diff
@@ -88,8 +151,33 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns input array.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null || array.Length == 0)
+            {
+                if (array != null && array.Length == 0 && index != 0)
```
Comment (ellmaster) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: I think this looks odd. There should be a better way of doing this.

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 6 above)
Comment (ellmaster) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: Yeah I just figured that I misunderstood your code. Sorry for that! It's fine. :)

**Comment by TelefunkenU87** [comment]
Thank you Ell and Kai for another review. I appreciate all the time and effort you two put into all of this!
